app/renderer/components/navigation/navigator.js

```diff
@@ -0,0 +1,342 @@
+const siteSettings = require('../../../../js/state/siteSettings')
+const siteUtil = require('../../../../js/state/siteUtil')
+const cx = require('../../../../js/lib/classSet')
+
+// Constants
+const settings = require('../../../../js/constants/settings')
+const appConfig = require('../../../../js/constants/appConfig')
+
+class Navigator extends ImmutableComponent {
+  constructor () {
+    super()
+    this.onBack = this.onBack.bind(this)
+    this.onForward = this.onForward.bind(this)
+    this.onBackLongPress = this.onBackLongPress.bind(this)
+    this.onForwardLongPress = this.onForwardLongPress.bind(this)
+    this.onDoubleClick = this.onDoubleClick.bind(this)
+    this.onDragOver = this.onDragOver.bind(this)
+    this.onDrop = this.onDrop.bind(this)
+    this.onBraveMenu = this.onBraveMenu.bind(this)
+  }
+
+  onNav (e, navCheckProp, navType, navAction) {
+    const activeFrame = frameStateUtil.getActiveFrame(this.props.windowState)
+    const activeTabId = tabUtil.activeTabId(this.props.windowState)
+    const activeTab = activeFrame ? this.props.appState.get('tabs').find((tab) => tab.get('tabId') === activeTabId) : null
+    const isNavigable = isNavigatableAboutPage(getBaseUrl(activeFrame.get('location')))
+    if (e && eventUtil.isForSecondaryAction(e) && isNavigable) {
+      if (activeTab && activeTab.get(navCheckProp)) {
+        appActions.tabCloned(activeTabId, {
+          [navType]: true,
+          active: !!e.shiftKey
+        })
+      }
+    } else {
+      navAction.call(this.activeFrame)
+    }
+  }
+
+  getTotalBlocks (frames) {
+    if (!frames) {
+      return false
+    }
+
+    frames = makeImmutable(frames)
+
+    const ads = frames.getIn(['adblock', 'blocked'])
+    const trackers = frames.getIn(['trackingProtection', 'blocked'])
+    const scripts = frames.getIn(['noScript', 'blocked'])
+    const fingerprint = frames.getIn(['fingerprintingProtection', 'blocked'])
+    const blocked = (ads && ads.size ? ads.size : 0) +
+      (trackers && trackers.size ? trackers.size : 0) +
+      (scripts && scripts.size ? scripts.size : 0) +
+      (fingerprint && fingerprint.size ? fingerprint.size : 0)
+
+    return (blocked === 0)
+      ? false
+      : ((blocked > 99)
+        ? '99+'
+        : blocked)
+  }
+
+  get extensionButtons () {
+    const activeTabId = tabUtil.activeTabId(this.props.windowState)
+    const enabledExtensions = extensionState.getEnabledExtensions(this.props.appState)
+    const extensionBrowserActions = enabledExtensions
+      .map((extension) => extensionState.getBrowserActionByTabId(this.props.appState, extension.get('id'), activeTabId))
+      .filter((browserAction) => browserAction)
+
+    let buttons = extensionBrowserActions.map((browserAction, id) =>
+      <BrowserAction
+        browserAction={browserAction}
+        extensionId={id}
+        tabId={activeTabId}
+        popupWindowSrc={this.props.windowState.getIn(['popupWindowDetail', 'src'])} />
```

Comment:
Can we pass in popupWindowSrc as a prop? See my general note below about passing in props for anything that uses this.props.windowState or this.props.appState. When we transition to ReduxComponent we will just move the props from the parent to `mergeProps`